File: 3488.closest-equal-element-queries.java
#
# @lc app=leetcode id=3488 lang=java
#
# [3488] Closest Equal Element Queries
#

import java.util.*;

# @lc code=start
class Solution {
    public List<Integer> solveQueries(int[] nums, int[] queries) {
        int n = nums.length;
        int[] minDist = new int[n];
        Arrays.fill(minDist, -1);
        Map<Integer, List<Integer>> mp = new HashMap<>();
        for (int i = 0; i < n; i++) {
            mp.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }
        for (Map.Entry<Integer, List<Integer>> entry : mp.entrySet()) {
            List<Integer> pos = entry.getValue();
            int k = pos.size();
            if (k < 2) continue;
            for (int i = 0; i < k; i++) {
                int p = pos.get(i);
                int lidx = (i - 1 + k) % k;
                int ridx = (i + 1) % k;
                int left = pos.get(lidx);
                int right = pos.get(ridx);
                int dl = Math.abs(p - left);
                dl = Math.min(dl, n - dl);
                int dr = Math.abs(p - right);
                dr = Math.min(dr, n - dr);
                minDist[p] = Math.min(dl, dr);
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int q : queries) {
            ans.add(minDist[q]);
        }
        return ans;
    }
}
# @lc code=end

File: 3488.closest-equal-element-queries.go
#
# @lc app=leetcode id=3488 lang=golang
#
# [3488] Closest Equal Element Queries
#

# @lc code=start
func solveQueries(nums []int, queries []int) []int {
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}
	n := len(nums)
	pos := make(map[int][]int)
	for i := range nums {
		pos[nums[i]] = append(pos[nums[i]], i)
	}
	ans := make([]int, n)
	for i := range ans {
		ans[i] = -1
	}
	for _, p := range pos {
		m := len(p)
		if m < 2 {
			continue
		}
		for k := 0; k < m; k++ {
			var left, right int
			if k == 0 {
				left = p[0] + n - p[m-1]
			} else {
				left = p[k] - p[k-1]
			}
			if k == m-1 {
				right = p[0] + n - p[m-1]
			} else {
				right = p[k+1] - p[k]
			}
			ans[p[k]] = min(left, right)
		}
	}
	result := make([]int, len(queries))
	for i, q := range queries {
		result[i] = ans[q]
	}
	return result
}
# @lc code=end